### ISSUE TITLE
Playfield objects appear to lag at the beginning of certain maps.

### ISSUE THREAD

**Issue by VisMajorSpeedruns** [ppy/osu#22488]
### Type

Performance

### Bug description

I've occasionally noticed a slight lag at the beginning of certain maps.
Looking at my footage frame by frame, my cursor still moves, but approach circles slow down or freeze briefly.
I've only noticed it on maps that have hit objects very close to the beginning of the map.
Some maps: 
https://osu.ppy.sh/beatmapsets/1129015#osu/2359641
https://osu.ppy.sh/beatmapsets/418202#osu/905342
https://osu.ppy.sh/beatmapsets/770023#osu/1618984
Harder to tell if it appears on this map as well, but I think it might.
https://osu.ppy.sh/beatmapsets/1095785#osu/2290176

### Screenshots or videos

https://user-images.githubusercontent.com/35517046/216144255-46948064-6319-40d9-a1fc-cd6d74737e1e.mp4

### Version

2023.123.0-lazer

### Logs

[performance.log](https://github.com/ppy/osu/files/10560958/performance.log)
[performance-audio.log](https://github.com/ppy/osu/files/10560959/performance-audio.log)
[performance-draw.log](https://github.com/ppy/osu/files/10560961/performance-draw.log)
[performance-input.log](https://github.com/ppy/osu/files/10560962/performance-input.log)
[performance-update.log](https://github.com/ppy/osu/files/10560963/performance-update.log)
[runtime.log](https://github.com/ppy/osu/files/10560964/runtime.log)
[updater.log](https://github.com/ppy/osu/files/10560965/updater.log)
[database.log](https://github.com/ppy/osu/files/10560966/database.log)
[input.log](https://github.com/ppy/osu/files/10560967/input.log)
[network.log](https://github.com/ppy/osu/files/10560968/network.log)

**Comment by Theighlin** [comment]
Same issue as described in https://github.com/ppy/osu/discussions/16515, and more visible in the video of the last reply https://www.youtube.com/watch?v=FNnIlO2X9sE

**Comment by peppy** [comment]
Probably best to leave this open if there's not already an issue tracking this.

Don't really have a solution in mind yet, beyond adding empty space to the start of the track as part of the byte stream (could work quite well, but would be a bit fiddly to implement).

For what it's worth, stable has the same issue for any beatmaps where the first hitobject is very close to the start of the audio track.

**Comment by peppy** [comment]
In the latest release, the effect of this should have been greatly reduced. If anyone was experiencing this, could you please check how it's feeling these days?

I do have a forward path to reduce this down to zero, but it requires a bit more work, so it would be good to know how we're doing to better prioritise the follow-up work. In local testing, things are feeling pretty good (still noticeable but not as jarring/distracting), but this could be per-system and per-paltform.

**Comment by Theighlin** [comment]
Stutter seems to be the same for me from the few beatmaps i've tried.

**Comment by Xerbly** [comment]
I can still recreate this issue on a few maps, in version 2023.1130.0. Also, switching audio outputs somehow fixes the issue temporarily. Here is a video:

https://github.com/ppy/osu/assets/89270842/cc4fc85f-aa27-4ee4-9273-07821bee105b

Also logs if you need them.
[performance-update.log](https://github.com/ppy/osu/files/13652110/performance-update.log)
[runtime.log](https://github.com/ppy/osu/files/13652111/runtime.log)
[input.log](https://github.com/ppy/osu/files/13652114/input.log)
[performance.log](https://github.com/ppy/osu/files/13652115/performance.log)
[performance-draw.log](https://github.com/ppy/osu/files/13652116/performance-draw.log)
[database.log](https://github.com/ppy/osu/files/13652112/database.log)

**Comment by Maxwell-lt** [comment]
I've been able to consistently reproduce what looks like the same issue on the latest release (2023.1223.0-lazer). I've recorded a few plays that have this issue, and in all cases recordings made with OBS show that there is a pause of about 133-150ms (9 total frames with the same position of approach circles/slider snaking, recorded at 60FPS). 

This happens consistently for each of the beatmaps I've observed this issue on, which includes all four of the beatmaps listed in the original issue description. However, I was not able to reproduce the issue on the beatmap @Xerbly used https://osu.ppy.sh/beatmapsets/1621390#osu/3310382. (Nor did switching audio outputs have any effect for me.)

This occurs in all four game modes, per my testing on https://osu.ppy.sh/beatmapsets/1530096#osu/3370765

My scenario appears to be more severe than what's been previously reported, as it causes the audio to play >100ms ahead of the gameplay timing (most easily observed when using the Auto mod), requiring the circles to be clicked very offbeat for an optimal judgement.

Once a break happens, the audio and gameplay regain sync.

I've tried windowed, borderless, and fullscreen screen modes, as well as multiple frame limiter settings (tried VSync, 4x, and unlimited), and both multi and single thread modes. This issue was reproducible in each of those scenarios.

This is not an issue that I can reproduce on osu!stable. I tried https://osu.ppy.sh/beatmapsets/1002234#osu/2097830 (see video below for behavior on osu!lazer) on stable and did not encounter a stutter or audio timing issues.

### System details

**Kernel**: Linux 6.1.65
**OS**: NixOS 24.05.20231204.2c7f3c0
**Desktop Environment**: KDE Plasma 5.27.9
**Window Manager**: KWin (Wayland)
**Audio Backend**: Pipewire
**CPU**: AMD Ryzen 9 7950X3D
**GPU**: AMD Radeon RX 6950 XT

### Videos

Video showing a stutter near start of approach circle on https://osu.ppy.sh/beatmapsets/1002234#osu/2097830
(at 00:09 in video)

https://github.com/ppy/osu/assets/17859747/6532a465-56f9-4f8c-88ee-7c03766958d8

Video showing a stutter near end of approach circle then audio sync becoming correct after a break on https://osu.ppy.sh/beatmapsets/1530096#osu/3435055
(at 00:06 in video)

https://github.com/ppy/osu/assets/17859747/68e00dff-1c26-48dd-a058-4cba7cdfde35

Video showing a stutter with the frame statistics overlay visible on https://osu.ppy.sh/beatmapsets/211586#osu/497556
(at 00:05 in video)

https://github.com/ppy/osu/assets/17859747/1143d621-230d-4c13-9326-335682135420

### Logs

Exported after a single play with this issue

[compressed-logs.zip](https://github.com/ppy/osu/files/13757421/compressed-logs.zip)

**Comment by peppy** [comment]
Thanks for the report, but we're already aware of how to repro this and we are just waiting on having enough time to fix it.

**Comment by Theighlin** [comment]
Heads up that https://github.com/ppy/osu-framework/pull/6088 fixes the issue for me (Tested on more than 1 map and even a couple from OP)
Release

https://github.com/ppy/osu/assets/53081750/73702ca0-999c-4e76-9753-e22e85d73a89

PR

https://github.com/ppy/osu/assets/53081750/25a12519-9f48-4f0b-b9bd-a4e0f4d60028

**Comment by peppy** [comment]
I suspect the top level audio mixer does heavy lifting here, and may solve the issue on other platforms when we do the same there (although it's arguably worse on window than anywhere else, so this is already a relief).

**Comment by mattborkin** [comment]
This happens to me with the lazy release when playing Osu! mania as well.

I believe I read above that the issue happens when the notes start right away without an empty space ahead of it. That being the case, could a quick fix be to simply somehow add blank space and silence before every track automatically?

**Comment by peppy** [comment]
> Leaving it as a reference for when a fix is explored for this, BASS seems to offer an option to delay the start of a track via [`BASS_Mixer_StreamAddChannelEx`](https://www.un4seen.com/doc/#bassmix/BASS_Mixer_StreamAddChannelEx.html). I'm not sure how reliable it is and whether it will help here but probably worth mentioning it at least.

This doesn't say anything about initialising the playback during that delay. It seems to be more about queueing it. Have you actually tested this to improve the issue here?

**Comment by frenzibyte** [comment]
I haven't, I only stated it for reference as I found it on passing. I didn't think it's something I can test on a whim.

**Comment by ProjWyrd** [comment]
I'm just going to add some info here which was in my issue but didn't realize it was a duplicate. 

I've found that this is still somewhat of an issue despite the many updates but there is a few things I've found.

1. The issue doesn't occur on the first run of opening lazer but rather after doing about 10 or so the issue becomes present and more pronounced
2. When the issue occured it seemed as if there was some *additional* offset on the sound resulting in a bugged and normal run sounding completely identical but the input timing to be way off. (look at the videos on #31497 to better understand)

**Comment by bad1080** [comment]
does "when lead-in time is used" mean there's a way around the stutters and if so how? or is that something map makers can opt to use when creating a beat-map?

**Comment by peppy** [comment]
A prospective fix for this is being rolled out in today's tachyon release. Please give it a try if this is still affecting you.

**Comment by Semro** [comment]
> A prospective fix for this is being rolled out in today's tachyon release. Please give it a try if this is still affecting you.

same behavior as described here: https://github.com/ppy/osu/issues/24985

version: 2025.827.0-tachyon

**Comment by peppy** [comment]
Does this mean it has improved for you?

**Comment by bad1080** [comment]
the issue has greatly improved for me with "2025.827.0-tachyon". on some maps the stutter is completely gone and on others it feels like it's down from like 400ms to maybe 150ms or so (i did no frame analysis or the like, this is purely from feeling).
(this is on mania btw)

**Comment by peppy** [comment]
The original stutter was ~9 ms, and now it should feel more like ~4 ms at most. Your time scale is off, but thanks for your feedback.

**Comment by Semro** [comment]
> Does this mean it has improved for you?

I think, it did.

To spot the difference, recorded videos on map with most noticeable stutter  -  [Turkish March - Owata (^o^)/](https://osu.ppy.sh/beatmapsets/399096#osu/1004492) and watched them frame by frame.

- lazer. In auto mod cursor was slowing down/speeding up in a span of ~41 ms.
- tachyon. Also in auto mod can't see cursor speed change, at least at 144fps.

I had another issue which were causing ~200ms stutter at the same timestamp on the beatmap and more likely related to my setup.
<details>
  <summary>Issue description, fix</summary>
  
This stutter was only appearing when using wireless headphones with A2DP audio profiles (SBC and LDAC codecs). 

Stutter disappeared when I changed speaker latency offset to -200ms.

If you are on Linux you can change it in pavucontrol: Output Devices/<your headphones>/Advanced/Latency Offset, for more info look into [this answer](https://askubuntu.com/questions/145935/get-rid-of-0-5s-latency-when-playing-audio-over-bluetooth-with-a2dp)

To play comfortably in osu! you should:
 - change global offset. Set it to your speaker latency offset and then adjust if needed blindly playing [Offset Wizard map](https://osu.ppy.sh/beatmapsets/4659#osu/24854) (you can tap on every bit), click on your score and look into Average Hit Error (for me there was additional -40ms)
 - set "Effect" to 0, not to be distracted by delayed hitsounds. Could there be an option just to disable only gameplay hitsounds and not all in game sounds, cause they are awesome?

</details>

**Comment by peppy** [comment]
Will be testing a smoother setting in the next tachyon release, so stay tuned for that.

**Comment by liesened** [comment]
On 2025.1029.1 when using Pipewire's Pro Audio in order to get rid of OS-level mixing and achieve near-zero audio latency, for some reason osu! completely freaks out and you have to use positive offset of +300 for the game to sync with the audio if the map starts later than the audio does and +224 if the map starts before the audio.

Needless to say you are getting a massive stutter in the beginning of the map in either case (similar to what's in the OP), and you even get it when you look at the map in the editor.

The hitsounds are indeed near-instantenuos, though. I use 128/48000 quant size (pw-top reports 117/44100 for osu!), and I almost can't hear any delay between the keyboard sound and a hitsound.

**Comment by bdach** [comment]
@liesened we do not provide support for custom tweaks to pipewire settings, you'll have to debug this yourself

**Comment by liesened** [comment]
I don't ask for a Pipewire support, this is an issue with how osu! handles audio, this has nothing to do with Pipewire by itself because Pro Audio profile works flawlessly in other regular applications and even in osu! as long as you set the audio offset to +300 (or +224, see above) and pretend the lag in the beginning of the map doesn't exist.

Besides, I don't think Pro Audio profile should be classified as "custom tweaks", it is a [well-documented](https://gitlab.freedesktop.org/pipewire/pipewire/-/wikis/FAQ#what-is-the-pro-audio-profile) feature and it allows achieving extremely low audio latency without doing much at all on all modern devices. It's weird osu! does not work well using this profile in the first place, as it's a rhythm game where you *do* want next-to-nothing audio latency.

**Comment by sophietelle** [comment]
updating from https://osu.ppy.sh/community/forums/topics/2163751

Problem details:
When default profile used, sound works as expected.
When Pro Mode profile is used, the song feels like it started too late (same goes for intro music by the way), and when the beatmap starts the approach circle freezes for +/- 300ms.
You can even notice it from the video - the map is horribly off-beat.
The strange thing is that hitsounds are not delayed at all - they play perfectly in time and with even lower latency than with default profile.

Video or screenshot showing the problem: https://youtu.be/G2QX_bTQPq8
Minimal reproduction logs (start osu!, go to a beatmap, start it, play 3 seconds, export logs): https://files.catbox.moe/kmu5zx.zip

+ when I ran osu! from source in debug I got this log at the start of the map:
[verbose]: Playback discrepancy detected (1 of allowed 5): 1,702.3 vs 2,002.3

osu! version: Lazer 2025.1218.0-lazer-linux (latest)

I've successfully reproduced this bug on a live Fedora install: https://www.youtube.com/watch?v=cuBVar6YuH0
The wpctl output if needed: https://files.catbox.moe/y73vru.zip

I am also free to test any changes and build lazer from source with any patches to test something, just hmu
